### PR TITLE
[Fixes #263] Tests CRUD of the Application - Read Projects

### LINKED ISSUE #263: Testing CRUD of the Application - Read Projects
## Feature Request

Is your feature request related to an existing [Issue](https://www.github.com/Auto-DL/Generator/issues) or a new implementation?

No

### Feature Description

Write tests for the Frontend (using React Testing library, etc.) to verify the "Read Projects" part of the CRUD of the entire Application. Isolate the testing to the frontend by creating mocks or using any other suitable solution, so as to prevent testing the Backend APIs as well.

### PR DESCRIPTION
**Pull request by McTechie**
# Pull Request

## What does this PR do?
Fixes #263

The following are the ways in which the tests have been catered to:
- Isolated testing to the Frontend (tests do not verify the correctness of the Backend APIs).
- Prevented mocking axios requests in order to test like an actual user, as well as reduce requests cost to the server.
- Configured **Mock Service Worker** (MSW) to intervene axios requests.
- Created mocks and handlers for "Read Projects" part of the CRUD of the app.

> Updated package.json -> Added msw as **development dependency**

## What part does this affect?
- [x] FrontEnd.
- [ ] BackEnd.
- [ ] Documentation.
- [ ] Other. (Please specify below)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue or slack?
- [x] Did you read the [contributor guideline](https://github.com/Auto-DL/Auto-DL/blob/v1-beta/CONTRIBUTING.md)?
- [x] Did you ensure that there aren't any other open [Pull Requests](https://www.github.com/Auto-DL/Generator/pulls) for the same update/change?
- [x] Did you make sure the title is self-explanatory and the description concisely explains the PR?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure the code is clean and docstrings have been added or updated as required?
- [x] Did you make sure the code is linted/formatted locally prior to submission? (using `black` and/or `prettier`)
- [ ] Did you make sure to update the documentation with your changes? (if necessary)

## PR review
Anyone in the community is free to review the PR once the tests have passed.

Thank you for contributing to [AutoDL](https://github.com/Auto-DL/Auto-DL). We look forward to your continued support.

### CODE REVIEW

File: FrontEndApp/v1-react/src/mocks/resolvers/readProjectsResolver.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    output_file_name: "test.py",
+    username: "testUser",
+};
+
+export const getMockProjects = (req, res, ctx) => {
+    return res(
+        ctx.status(200),
+        ctx.json({ projects: mockResponse }),
+    )
+}
```
Comment (ADI10HERO) [inline FrontEndApp/v1-react/src/mocks/resolvers/readProjectsResolver.js:22]:
```suggestion
}

```

File: FrontEndApp/v1-react/src/mocks/resolvers/createProjectsResolver.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    output_file_name: "test.py",
+    username: "testUser",
+};
+
+export const createMockProject = (req, res, ctx) => {
+    return res(
+        ctx.status(200),
+        ctx.json({ projects: mockResponse }),
+    )
+}
```
Comment (ADI10HERO) [inline FrontEndApp/v1-react/src/mocks/resolvers/createProjectsResolver.js:22]:
```suggestion
}

```

**Review by ADI10HERO — CHANGES_REQUESTED** [review]

File: FrontEndApp/v1-react/src/mocks/handlers.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { rest } from "msw";
+import { getMockProjects } from "./resolvers/readProjectsResolver";
+import { createMockProject } from "./resolvers/createProjectsResolver";
+
+const handlers = [
+    rest.post("http://localhost/v1/projects/all/", getMockProjects),
+    rest.post("http://localhost/v1/project/new/", createMockProject),
+];  
+
+export { handlers, rest };
```
[the remark this answers is not in the document]
Comment (McTechie) [reply FrontEndApp/v1-react/src/mocks/handlers.js]: Yes... I'll do the same xD

File: FrontEndApp/v1-react/src/Components/MainApp/__test__/Homeindex.test.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+// describe("Create Projects", () => {
+//     test("should create a new project", async () => {
+//         render(<Homeindex />);
+
+//         await waitFor(() => {
+//             let btnElement = screen.getByTestId("project-save-button");
+//             expect(btnElement).toBeInTheDocument();
+//         });
+//     });
+// });
```
[the remark this answers is not in the document]
Comment (McTechie) [reply FrontEndApp/v1-react/src/Components/MainApp/__test__/Homeindex.test.js]: Ahh it was for the next CRUD operation... but i've removed the code (will add it from scratch for the next pull request)

**Comment by McTechie** [comment]
@ADI10HERO @rajraj889 I've made the suggested changes... and will be creating new PRs on the other CRUD operations as well

**Review by ADI10HERO — APPROVED** [review]
@McTechie  The tests are passing but can you have a look into why this is happening when I run tests?

image

I will create an issue for this if you feel its an issue

**Comment by McTechie** [comment]
@ADI10HERO thanks for pointing this out... I have verified that two of these console errors were appearing because of a few bugs in our frontend (ProjectTable.js) code... I have rectified these and now the tests should pass without any console errors.
However, there are a few more errors and warnings onto the console of the frontend.... so I'll try to create an issue regarding fixing the same.

**Comment by ADI10HERO** [comment]
I am still getting the `console.error` when running all tests please create the issue.

**Pull request merged**